File: pirates-android/app/src/main/java/com/crescentflare/piratesgame/components/containers/FrameContainerView.kt
package com.crescentflare.piratesgame.components.containers

import android.annotation.TargetApi
import android.content.Context
import android.os.Build
import android.util.AttributeSet
import com.crescentflare.piratesgame.components.utility.ViewletUtil
import com.crescentflare.piratesgame.infrastructure.events.AppEvent
import com.crescentflare.piratesgame.infrastructure.events.AppEventLabeledSender
import com.crescentflare.piratesgame.infrastructure.events.AppEventObserver
import com.crescentflare.unilayout.containers.UniFrameContainer
import com.crescentflare.jsoninflator.JsonInflatable
import com.crescentflare.jsoninflator.binder.InflatorBinder
import com.crescentflare.jsoninflator.utility.InflatorMapUtil
import java.lang.ref.WeakReference


/**
 * Container view: basic layout container for overlapping views
 */
open class FrameContainerView : UniFrameContainer, AppEventObserver, AppEventLabeledSender {

    // --
    // Static: viewlet integration
    // --

    companion object {

        val viewlet: JsonInflatable = object : JsonInflatable {
            override fun create(context: Context): Any {
                return FrameContainerView(context)
            }

            override fun update(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>, parent: Any?, binder: InflatorBinder?): Boolean {
                if (obj is FrameContainerView) {
                    // Set container name
                    obj.contentDescription = mapUtil.optionalString(attributes, "containerName", null)

                    // Create or update children
                    ViewletUtil.createSubviews(mapUtil, obj, obj, attributes, attributes["items"], binder)

                    // Generic view properties
                    ViewletUtil.applyGenericViewAttributes(mapUtil, obj, attributes)

                    // Event handling
                    obj.tapEvent = AppEvent.fromObject(attributes["tapEvent"])

                    // Forward event observer
                    if (parent is AppEventObserver) {
                        obj.eventObserver = parent
                    }
                    return true
                }
                return false
            }

            override fun canRecycle(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>): Boolean {
                return obj is FrameContainerView
            }
        }
    }


    // --
    // Members
    // --

    private var eventObserverReference : WeakReference<AppEventObserver>? = null


    // --
    // Initialization
    // --

    @JvmOverloads
    constructor(
            context: Context,
            attrs: AttributeSet? = null,
            defStyleAttr: Int = 0)
            : super(context, attrs, defStyleAttr)

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(
            context: Context,
            attrs: AttributeSet?,
            defStyleAttr: Int,
            defStyleRes: Int)
            : super(context, attrs, defStyleAttr, defStyleRes)

    init {
        // No implementation
    }


    // --
    // Configurable values
    // --

    var eventObserver: AppEventObserver?
        get() = eventObserverReference?.get()
        set(newValue) {
            eventObserverReference = if (newValue != null) {
                WeakReference(newValue)
            } else {
                null
            }
        }

    var tapEvent: AppEvent? = null
        set(tapEvent) {
            field = tapEvent
            if (hasOnClickListeners()) {
                setOnClickListener(null)
            }
            if (this.tapEvent != null) {
                setOnClickListener {
                    val currentTapEvent = this@FrameContainerView.tapEvent
                    if (currentTapEvent != null) {
                        observedEvent(currentTapEvent, this@FrameContainerView)
                    }
                }
            }
        }


    // --
    // Interaction
    // --

    override val senderLabel: String?
        get() = contentDescription?.toString()

    override fun observedEvent(event: AppEvent, sender: Any?) {
        eventObserver?.observedEvent(event, sender)
    }

}

File: pirates-ios/PiratesGame/Components/Containers/FrameContainerView.swift
//
//  FrameContainerView.swift
//  Container view: basic layout container for overlapping views
//

import UIKit
import UniLayout
import JsonInflator

class FrameContainerView: UniFrameContainer, AppEventObserver, AppEventLabeledSender, UniTapDelegate {
    
    // --
    // MARK: Viewlet integration
    // --
    
    class func viewlet() -> JsonInflatable {
        return ViewletClass()
    }
    
    private class ViewletClass: JsonInflatable {
        
        func create() -> Any {
            return FrameContainerView()
        }
        
        func update(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any], parent: Any?, binder: InflatorBinder?) -> Bool {
            if let frameContainer = object as? FrameContainerView {
                // Set container name
                frameContainer.accessibilityLabel = convUtil.asString(value: attributes["containerName"])
                
                // Create or update subviews
                ViewletUtil.createSubviews(convUtil: convUtil, container: frameContainer, parent: frameContainer, attributes: attributes, subviewItems: attributes["items"], binder: binder)
                
                // Generic view properties
                ViewletUtil.applyGenericViewAttributes(convUtil: convUtil, view: frameContainer, attributes: attributes)
                
                // Event handling
                frameContainer.tapEvent = AppEvent(value: attributes["tapEvent"])
                
                // Forward event observer
                if let eventObserver = parent as? AppEventObserver {
                    frameContainer.eventObserver = eventObserver
                }
                return true
            }
            return false
        }
        
        func canRecycle(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any]) -> Bool {
            return object is FrameContainerView
        }
        
    }
    
    
    // --
    // MARK: Initialization
    // --
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }
    
    private func setup() {
    }
    
    
    // --
    // MARK: Configurable values
    // --

    weak var eventObserver: AppEventObserver?

    var tapEvent: AppEvent? {
        didSet {
            tapDelegate = tapEvent != nil ? self : nil
        }
    }
    
    
    // --
    // MARK: Interaction
    // --
    
    var senderLabel: String? {
        get {
            return accessibilityLabel
        }
    }
    
    func containerTapped(_ sender: UIView) {
        if let tapEvent = tapEvent {
            observedEvent(tapEvent, sender: self)
        }
    }
    
    func observedEvent(_ event: AppEvent, sender: Any?) {
        eventObserver?.observedEvent(event, sender: sender)
    }
    
}
